test/Microsoft.AspNetCore.Mvc.Razor.Test/Internal/DefaultRoslynCompilationServiceTest.cs

```diff
@@ -0,0 +1,313 @@
+        {
+            // Arrange
+            var content = @"
+public class MyTestType  {}";
+
+            var compilationService = new DefaultRoslynCompilationService(
+                GetDependencyContext(),
+                GetOptions(),
+                GetFileProviderAccessor(),
+                NullLoggerFactory.Instance);
+            var relativeFileInfo = new RelativeFileInfo(
+                new TestFileInfo { PhysicalPath = "SomePath" },
+                "some-relative-path");
+
+            // Act
+            var result = compilationService.Compile(relativeFileInfo, content);
+
+            // Assert
+            Assert.Equal("MyTestType", result.CompiledType.Name);
+        }
+
+        [Fact]
+        public void Compile_ReturnsCompilationFailureWithPathsFromLinePragmas()
+        {
+            // Arrange
+            var viewPath = "some-relative-path";
+            var fileContent = "test file content";
+            var content = $@"
+#line 1 ""{viewPath}""
+this should fail";
+            var fileProvider = new TestFileProvider();
+            var fileInfo = fileProvider.AddFile(viewPath, fileContent);
+
+            var compilationService = new DefaultRoslynCompilationService(
+                GetDependencyContext(),
+                GetOptions(),
+                GetFileProviderAccessor(fileProvider),
+                NullLoggerFactory.Instance);
+            var relativeFileInfo = new RelativeFileInfo(fileInfo, "some-relative-path");
+
+            // Act
+            var result = compilationService.Compile(relativeFileInfo, content);
+
+            // Assert
+            Assert.IsType<CompilationResult>(result);
+            Assert.Null(result.CompiledType);
+            var compilationFailure = Assert.Single(result.CompilationFailures);
+            Assert.Equal(relativeFileInfo.RelativePath, compilationFailure.SourceFilePath);
+            Assert.Equal(fileContent, compilationFailure.SourceFileContent);
+        }
+
+        [Fact]
+        public void Compile_ReturnsGeneratedCodePath_IfLinePragmaIsNotAvailable()
+        {
+            // Arrange
+            var fileContent = "file content";
+            var content = @"this should fail";
+
+            var compilationService = new DefaultRoslynCompilationService(
+                GetDependencyContext(),
+                GetOptions(),
+                GetFileProviderAccessor(),
+                NullLoggerFactory.Instance);
+            var relativeFileInfo = new RelativeFileInfo(
+                new TestFileInfo { Content = fileContent },
+                "some-relative-path");
+
+            // Act
+            var result = compilationService.Compile(relativeFileInfo, content);
+
+            // Assert
+            Assert.IsType<CompilationResult>(result);
+            Assert.Null(result.CompiledType);
+
+            var compilationFailure = Assert.Single(result.CompilationFailures);
+            Assert.Equal("Generated Code", compilationFailure.SourceFilePath);
```

Comment:
This is some quirky behavior related to - https://github.com/dotnet/roslyn/issues/1920. If we return the `SourceFilePath` here, it ends up not allowing breakpoints to be set unless you have the Allow source files to be different... flag set in the VS.